mimic/rest/auth_api.py

```diff
@@ -44,41 +45,79 @@ def get_token_and_service_catalog(self, request):
+
+        def format_response(callable_returning_session,
+                            nonmatching_tenant_message_generator):
+            try:
+                session = callable_returning_session()
+            except NonMatchingTenantError as e:
+                request.setResponseCode(401)
+                return json.dumps({
+                    "unauthorized": {
+                        "code": 401,
+                        "message": nonmatching_tenant_message_generator(e)
+                    }
+                })
+            else:
+                request.setResponseCode(200)
+                prefix_map = {
+                    # map of entry to URI prefix for that entry
+                }
+
+                def lookup(entry):
+                    return prefix_map[entry]
+                result = get_token(
+                    session.tenant_id,
+                    entry_generator=lambda tenant_id:
+                    list(self.core.entries_for_tenant(
+                         tenant_id, prefix_map, base_uri_from_request(request))),
+                    prefix_for_endpoint=lookup,
+                    response_token=session.token,
+                    response_user_id=session.user_id,
+                    response_user_name=session.username,
+                )
+                result['access']['user']['roles'].append({
+                    'description': 'User Admin Role.',
+                    'id': '3',
+                    'name': 'identity:user-admin'})
+                return json.dumps(result)
+
+        username_generator = (
+            lambda exception: "Tenant with Name/Id: '{0}' is not valid for "
+                              "User '{1}' (id: '{2}')".format(
+                                  exception.desired_tenant,
+                                  exception.session.username,
+                                  exception.session.user_id))
+
         if content['auth'].get('passwordCredentials'):
             username = content['auth']['passwordCredentials']['username']
             password = content['auth']['passwordCredentials']['password']
-            session = self.core.sessions.session_for_username_password(
-                username, password, tenant_id)
+            return format_response(
+                lambda: self.core.sessions.session_for_username_password(
+                    username, password, tenant_id),
+                username_generator)
+
         elif content['auth'].get('RAX-KSKEY:apiKeyCredentials'):
-            username = content['auth']['RAX-KSKEY:apiKeyCredentials']['username']
-            api_key = content['auth']['RAX-KSKEY:apiKeyCredentials']['apiKey']
-            session = self.core.sessions.session_for_api_key(
-                username, api_key, tenant_id)
+            username = content['auth']['RAX-KSKEY:apiKeyCredentials'][
+                'username']
+            api_key = content['auth']['RAX-KSKEY:apiKeyCredentials'][
+                'apiKey']
```

Comment:
Come to think of it, maybe this is a place that we can split things up - now that this is indented so far that the repetition is causing every line to be wrapped, perhaps we could split up